File: packages/sdks/client-sdks/gatrix-flutter-sdk/lib/src/streaming_sse.dart
// SSE (Server-Sent Events) streaming connection for real-time flag updates.
// Uses dart:io HttpClient for chunked streaming with automatic reconnection.

import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';

import 'events.dart';
import 'models.dart';

/// Callback when flags are invalidated via streaming
typedef InvalidationCallback = void Function(List<String> changedKeys);

/// Callback for full-fetch request (gap recovery)
typedef FetchCallback = void Function();

class SseConnection {
  final String apiUrl;
  final String apiToken;
  final String appName;
  final String environment;
  final String connectionId;
  final String sdkVersion;
  final SseStreamingConfig config;
  final EventEmitter events;
  final Map<String, String>? customHeaders;

  StreamingConnectionState _state = StreamingConnectionState.disconnected;
  int _reconnectAttempt = 0;
  int _reconnectCount = 0;
  int _eventCount = 0;
  int _errorCount = 0;
  int _recoveryCount = 0;
  int _localGlobalRevision = 0;
  String? _lastError;
  DateTime? _lastEventTime;
  DateTime? _lastErrorTime;
  DateTime? _lastRecoveryTime;
  bool _stopRequested = false;
  Timer? _reconnectTimer;
  HttpClient? _httpClient;
  final Random _random = Random();

  InvalidationCallback? onInvalidation;
  FetchCallback? onFetchRequest;

  // SSE parser state
  String _currentEventType = '';
  StringBuffer _dataBuffer = StringBuffer();

  SseConnection({
    required this.apiUrl,
    required this.apiToken,
    required this.appName,
    required this.environment,
    required this.connectionId,
    required this.sdkVersion,
    required this.config,
    required this.events,
    this.customHeaders,
  });

  StreamingConnectionState get state => _state;
  int get reconnectCount => _reconnectCount;
  int get eventCount => _eventCount;
  int get errorCount => _errorCount;
  int get recoveryCount => _recoveryCount;
  String? get lastError => _lastError;
  DateTime? get lastEventTime => _lastEventTime;
  DateTime? get lastErrorTime => _lastErrorTime;
  DateTime? get lastRecoveryTime => _lastRecoveryTime;

  /// Start SSE connection
  void connect() {
    if (_state == StreamingConnectionState.connected ||
        _state == StreamingConnectionState.connecting) {
      return;
    }

    _state = StreamingConnectionState.connecting;
    _stopRequested = false;
    _runSseLoop();
  }

  /// Disconnect and cleanup
  void disconnect() {
    _stopRequested = true;
    _state = StreamingConnectionState.disconnected;
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
    _httpClient?.close(force: true);
    _httpClient = null;
  }

  String _buildUrl() {
    final baseUrl = config.url ??
        '$apiUrl/client/features/$environment/stream/sse';
    final params = <String, String>{
      'x-api-token': apiToken,
      'appName': appName,
      'environment': environment,
      'connectionId': connectionId,
      'sdkVersion': sdkVersion,
    };
    final uri = Uri.parse(baseUrl);
    return uri.replace(queryParameters: {...uri.queryParameters, ...params}).toString();
  }

  Future<void> _runSseLoop() async {
    final url = _buildUrl();

    try {
      _httpClient?.close(force: true);
      _httpClient = HttpClient();

      final request = await _httpClient!.getUrl(Uri.parse(url));
      request.headers.set('Accept', 'text/event-stream');
      request.headers.set('Cache-Control', 'no-cache');
      request.headers.set('X-API-Token', apiToken);
      request.headers.set('X-Application-Name', appName);
      request.headers.set('X-Environment', environment);
      request.headers.set('X-Connection-Id', connectionId);
      request.headers.set('X-SDK-Version', sdkVersion);
      customHeaders?.forEach((key, value) {
        request.headers.set(key, value);
      });

      final response = await request.close();

      if (response.statusCode != 200) {
        final errorMsg = 'SSE connection failed: HTTP ${response.statusCode}';
        _trackError(errorMsg);
        _state = StreamingConnectionState.reconnecting;
        events.emit(GatrixEvents.flagsStreamingError, [errorMsg]);
        events.emit(GatrixEvents.flagsStreamingDisconnected);
        _scheduleReconnect();
        return;
      }

      // Track recovery
      if (_reconnectCount > 0) {
        _trackRecovery();
      }
      _state = StreamingConnectionState.connected;
      _reconnectAttempt = 0;
      events.emit(GatrixEvents.flagsStreamingConnected);

      // Reset parser state
      _currentEventType = '';
      _dataBuffer = StringBuffer();

      // Listen to stream
      await for (final chunk in response.transform(utf8.decoder)) {
        if (_stopRequested) break;
        _parseSseChunk(chunk);
      }

      // Stream ended
      if (!_stopRequested &&
          _state != StreamingConnectionState.disconnected) {
        _state = StreamingConnectionState.reconnecting;
        events.emit(GatrixEvents.flagsStreamingDisconnected);
        _scheduleReconnect();
      }
    } catch (e) {
      if (_stopRequested || _state == StreamingConnectionState.disconnected) {
        return;
      }

      _trackError(e.toString());
      events.emit(GatrixEvents.flagsStreamingError, [e.toString()]);

      if (_state != StreamingConnectionState.reconnecting) {
        _state = StreamingConnectionState.reconnecting;
        events.emit(GatrixEvents.flagsStreamingDisconnected);
      }
      _scheduleReconnect();
    }
  }

  void _parseSseChunk(String chunk) {
    final lines = chunk.split('\n');
    for (final rawLine in lines) {
      final line = rawLine.trimRight();

      if (line.isEmpty) {
        // Empty line = dispatch event
        if (_currentEventType.isNotEmpty || _dataBuffer.isNotEmpty) {
          final eventType =
              _currentEventType.isEmpty ? 'message' : _currentEventType;
          final eventData = _dataBuffer.toString();
          _currentEventType = '';
          _dataBuffer = StringBuffer();
          _processEvent(eventType, eventData);
        }
      } else if (line.startsWith('event:')) {
        _currentEventType = line.substring(6).trimLeft();
      } else if (line.startsWith('data:')) {
        if (_dataBuffer.isNotEmpty) _dataBuffer.write('\n');
        _dataBuffer.write(line.substring(5).trimLeft());
      }
      // Ignore 'id:', 'retry:', and comment lines
    }
  }

  void _processEvent(String eventType, String eventData) {
    _lastEventTime = DateTime.now();
    _eventCount++;

    switch (eventType) {
      case 'connected':
        try {
          final data = jsonDecode(eventData) as Map<String, dynamic>;
          final serverRevision =
              (data['globalRevision'] as num?)?.toInt() ?? 0;

          if (serverRevision > _localGlobalRevision &&
              _localGlobalRevision > 0) {
            _localGlobalRevision = serverRevision;
            onFetchRequest?.call();
          } else if (_localGlobalRevision == 0) {
            _localGlobalRevision = serverRevision;
          }
        } catch (_) {}
        break;

      case 'flags_changed':
        try {
          final data = jsonDecode(eventData) as Map<String, dynamic>;
          final serverRevision =
              (data['globalRevision'] as num?)?.toInt() ?? 0;
          final changedKeys = (data['changedKeys'] as List<dynamic>?)
                  ?.map((k) => k.toString())
                  .toList() ??
              [];

          if (serverRevision > _localGlobalRevision) {
            _localGlobalRevision = serverRevision;
            events.emit(GatrixEvents.flagsInvalidated);
            onInvalidation?.call(changedKeys);
          }
        } catch (_) {}
        break;

      case 'heartbeat':
        break;
    }
  }

  void _scheduleReconnect() {
    if (_state == StreamingConnectionState.disconnected || _stopRequested) {
      return;
    }

    _reconnectTimer?.cancel();
    _reconnectAttempt++;
    _reconnectCount++;

    final baseMs = config.reconnectBase * 1000;
    final maxMs = config.reconnectMax * 1000;
    final exponentialDelay =
        min(baseMs * pow(2, _reconnectAttempt - 1).toInt(), maxMs);
    final jitter = _random.nextInt(1000);
    final delayMs = exponentialDelay + jitter;

    events.emit(GatrixEvents.flagsStreamingReconnecting);

    // Transition to degraded after several failed attempts
    if (_reconnectAttempt >= 5 &&
        _state != StreamingConnectionState.degraded) {
      _state = StreamingConnectionState.degraded;
    }

    _reconnectTimer = Timer(
      Duration(milliseconds: delayMs),
      () {
        if (!_stopRequested && _state != StreamingConnectionState.disconnected) {
          connect();
        }
      },
    );
  }

  void _trackError(String errorMessage) {
    _errorCount++;
    _lastErrorTime = DateTime.now();
    _lastError = errorMessage;
  }

  void _trackRecovery() {
    _recoveryCount++;
    _lastRecoveryTime = DateTime.now();
  }
}

File: packages/sdks/client-sdks/gatrix-python-client-sdk/gatrix/streaming_sse.py
"""SSE (Server-Sent Events) streaming connection for real-time flag updates.

Uses Python's urllib (stdlib) for chunked HTTP streaming with automatic
reconnection via exponential backoff + jitter.
"""
from __future__ import annotations

import json
import logging
import math
import random
import threading
import time
from datetime import datetime, timezone
from typing import Callable, Dict, List, Optional
from urllib.error import URLError
from urllib.request import Request, urlopen

from gatrix.events import EVENTS, EventEmitter
from gatrix.types import SseStreamingConfig, StreamingConnectionState

logger = logging.getLogger("gatrix.streaming.sse")

# Callbacks
InvalidationCallback = Callable[[List[str]], None]
FetchCallback = Callable[[], None]


class SseConnection:
    """SSE streaming connection with automatic reconnection."""

    def __init__(
        self,
        api_url: str,
        api_token: str,
        app_name: str,
        environment: str,
        connection_id: str,
        sdk_version: str,
        config: SseStreamingConfig,
        emitter: EventEmitter,
        custom_headers: Optional[Dict[str, str]] = None,
    ) -> None:
        self._api_url = api_url
        self._api_token = api_token
        self._app_name = app_name
        self._environment = environment
        self._connection_id = connection_id
        self._sdk_version = sdk_version
        self._config = config
        self._emitter = emitter
        self._custom_headers = custom_headers or {}

        self._state: StreamingConnectionState = "disconnected"
        self._reconnect_attempt = 0
        self._reconnect_count = 0
        self._event_count = 0
        self._error_count = 0
        self._recovery_count = 0
        self._local_global_revision = 0
        self._last_error: Optional[str] = None
        self._last_event_time: Optional[datetime] = None
        self._last_error_time: Optional[datetime] = None
        self._last_recovery_time: Optional[datetime] = None
        self._stop_requested = False
        self._thread: Optional[threading.Thread] = None
        self._reconnect_timer: Optional[threading.Timer] = None

        self.on_invalidation: Optional[InvalidationCallback] = None
        self.on_fetch_request: Optional[FetchCallback] = None

    @property
    def state(self) -> StreamingConnectionState:
        return self._state

    @property
    def reconnect_count(self) -> int:
        return self._reconnect_count

    @property
    def event_count(self) -> int:
        return self._event_count

    @property
    def error_count(self) -> int:
        return self._error_count

    @property
    def recovery_count(self) -> int:
        return self._recovery_count

    @property
    def last_error(self) -> Optional[str]:
        return self._last_error

    @property
    def last_event_time(self) -> Optional[datetime]:
        return self._last_event_time

    @property
    def last_error_time(self) -> Optional[datetime]:
        return self._last_error_time

    @property
    def last_recovery_time(self) -> Optional[datetime]:
        return self._last_recovery_time

    def connect(self) -> None:
        """Start SSE connection."""
        if self._state in ("connected", "connecting"):
            return

        self._state = "connecting"
        self._stop_requested = False

        self._thread = threading.Thread(
            target=self._run_sse_loop, daemon=True, name="gatrix-sse"
        )
        self._thread.start()

    def disconnect(self) -> None:
        """Disconnect and cleanup."""
        self._stop_requested = True
        self._state = "disconnected"
        if self._reconnect_timer:
            self._reconnect_timer.cancel()
            self._reconnect_timer = None

    def _build_url(self) -> str:
        base_url = self._config.url or (
            f"{self._api_url}/client/features/{self._environment}/stream/sse"
        )
        params = (
            f"x-api-token={self._api_token}"
            f"&appName={self._app_name}"
            f"&environment={self._environment}"
            f"&connectionId={self._connection_id}"
            f"&sdkVersion={self._sdk_version}"
        )
        sep = "&" if "?" in base_url else "?"
        return f"{base_url}{sep}{params}"

    def _run_sse_loop(self) -> None:
        url = self._build_url()
        logger.debug("SSE connecting to: %s", url)

        try:
            req = Request(url)
            req.add_header("Accept", "text/event-stream")
            req.add_header("Cache-Control", "no-cache")
            req.add_header("X-API-Token", self._api_token)
            req.add_header("X-Application-Name", self._app_name)
            req.add_header("X-Environment", self._environment)
            req.add_header("X-Connection-Id", self._connection_id)
            req.add_header("X-SDK-Version", self._sdk_version)
            for k, v in self._custom_headers.items():
                req.add_header(k, v)

            with urlopen(req, timeout=None) as response:
                if response.status != 200:
                    error_msg = f"SSE HTTP error: {response.status}"
                    self._track_error(error_msg)
                    self._state = "reconnecting"
                    self._emitter.emit(EVENTS.FLAGS_STREAMING_ERROR, error_msg)
                    self._emitter.emit(EVENTS.FLAGS_STREAMING_DISCONNECTED)
                    self._schedule_reconnect()
                    return

                # Track recovery
                if self._reconnect_count > 0:
                    self._track_recovery()
                self._state = "connected"
                self._reconnect_attempt = 0
                logger.info("SSE streaming connected")
                self._emitter.emit(EVENTS.FLAGS_STREAMING_CONNECTED)

                # Parse SSE stream line by line
                current_event_type = ""
                data_buffer = ""
                for raw_line in response:
                    if self._stop_requested:
                        break

                    line = raw_line.decode("utf-8", errors="replace").rstrip(
                        "\r\n"
                    )

                    if not line:
                        # Empty line = dispatch
                        if current_event_type or data_buffer:
                            event_type = current_event_type or "message"
                            self._process_event(event_type, data_buffer)
                            current_event_type = ""
                            data_buffer = ""
                    elif line.startswith("event:"):
                        current_event_type = line[6:].lstrip()
                    elif line.startswith("data:"):
                        data_value = line[5:].lstrip()
                        if data_buffer:
                            data_buffer += "\n"
                        data_buffer += data_value
                    # Ignore id:, retry:, and comment lines

            # Stream ended
            if (
                not self._stop_requested
                and self._state != "disconnected"
            ):
                logger.info("SSE connection closed by server")
                self._state = "reconnecting"
                self._emitter.emit(EVENTS.FLAGS_STREAMING_DISCONNECTED)
                self._schedule_reconnect()

        except (URLError, OSError, Exception) as e:
            if self._stop_requested or self._state == "disconnected":
                return

            error_msg = str(e)
            self._track_error(error_msg)
            logger.warning("SSE connection error: %s", error_msg)
            self._emitter.emit(EVENTS.FLAGS_STREAMING_ERROR, error_msg)

            if self._state != "reconnecting":
                self._state = "reconnecting"
                self._emitter.emit(EVENTS.FLAGS_STREAMING_DISCONNECTED)
            self._schedule_reconnect()

    def _process_event(self, event_type: str, event_data: str) -> None:
        self._last_event_time = datetime.now(timezone.utc)
        self._event_count += 1

        if event_type == "connected":
            try:
                data = json.loads(event_data)
                server_revision = int(data.get("globalRevision", 0))
                if (
                    server_revision > self._local_global_revision
                    and self._local_global_revision > 0
                ):
                    self._local_global_revision = server_revision
                    if self.on_fetch_request:
                        self.on_fetch_request()
                elif self._local_global_revision == 0:
                    self._local_global_revision = server_revision
            except (json.JSONDecodeError, ValueError, TypeError):
                pass

        elif event_type == "flags_changed":
            try:
                data = json.loads(event_data)
                server_revision = int(data.get("globalRevision", 0))
                changed_keys = [str(k) for k in data.get("changedKeys", [])]

                if server_revision > self._local_global_revision:
                    self._local_global_revision = server_revision
                    self._emitter.emit(EVENTS.FLAGS_INVALIDATED)
                    if self.on_invalidation:
                        self.on_invalidation(changed_keys)
            except (json.JSONDecodeError, ValueError, TypeError):
                pass

        elif event_type == "heartbeat":
            logger.debug("SSE heartbeat received")

    def _schedule_reconnect(self) -> None:
        if self._state == "disconnected" or self._stop_requested:
            return

        if self._reconnect_timer:
            self._reconnect_timer.cancel()

        self._reconnect_attempt += 1
        self._reconnect_count += 1

        base_s = self._config.reconnect_base
        max_s = self._config.reconnect_max
        exponential = min(
            base_s * math.pow(2, self._reconnect_attempt - 1), max_s
        )
        jitter = random.uniform(0, 1)
        delay = exponential + jitter

        logger.debug(
            "Scheduling SSE reconnect: attempt=%d, delay=%.1fs",
            self._reconnect_attempt,
            delay,
        )
        self._emitter.emit(EVENTS.FLAGS_STREAMING_RECONNECTING)

        if self._reconnect_attempt >= 5 and self._state != "degraded":
            self._state = "degraded"

        def _reconnect():
            if not self._stop_requested and self._state != "disconnected":
                self.connect()

        self._reconnect_timer = threading.Timer(delay, _reconnect)
        self._reconnect_timer.daemon = True
        self._reconnect_timer.start()

    def _track_error(self, error_message: str) -> None:
        self._error_count += 1
        self._last_error_time = datetime.now(timezone.utc)
        self._last_error = error_message

    def _track_recovery(self) -> None:
        self._recovery_count += 1
        self._last_recovery_time = datetime.now(timezone.utc)
